File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/Teleop.java
package org.firstinspires.ftc.teamcode.teleop;

import static java.lang.Thread.sleep;

import com.acmerobotics.dashboard.config.Config;
import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.util.ElapsedTime;

import org.firstinspires.ftc.teamcode.backupAuton.auton.AutonMethods;
import org.firstinspires.ftc.teamcode.utilities.ClawRotator;
import org.firstinspires.ftc.teamcode.utilities.Slides;
import org.firstinspires.ftc.teamcode.utilities.Claw;
import org.firstinspires.ftc.teamcode.utilities.PullUp;
import org.firstinspires.ftc.teamcode.utilities.SimpleMecanumDrive;
import org.firstinspires.ftc.teamcode.utilities.SlidesState;

@Config
@TeleOp(name="DriveTrain Teleop")
public class Teleop extends OpMode {
    Slides slides;
    PullUp pullup;
    Claw claw;
    ClawRotator clawRotator;
    SimpleMecanumDrive drive;

    final double normalPower = 0.9;

    // in case of joystick drift, ignore very small values
    public float stick_margin = 0.7f;

    public boolean clawOpen = true;
    public boolean cRotatorAtDrop = false;

    enum ButtonPressState {
        PRESSED_GOOD, //the first time we see the button pressed
        DEPRESSED, //you haven't let go
        UNPRESSED // its not pressed
    }
    ButtonPressState slideButton;
    ButtonPressState rotatorButton;
    ButtonPressState clawButton;
    ButtonPressState cRotatorButton;
    ButtonPressState slideManual;
    ButtonPressState slideManualUp;
    ButtonPressState ultimateButton;

    boolean slowMode = false;
    boolean pullupstate1 = false;
    boolean pullupstate2 = false;
    boolean pulldownstate1 = false;
    boolean pulldownstate2 = false;
    int slidesPos;
    double time;
    int autonAction = 0;

    ElapsedTime runtime = new ElapsedTime();
    @Override
    public void init() {
        this.drive = new SimpleMecanumDrive(hardwareMap);
        this.slides = new Slides(hardwareMap);
        this.pullup = new PullUp(hardwareMap);
        this.clawButton = ButtonPressState.UNPRESSED;
        this.cRotatorButton = ButtonPressState.UNPRESSED;
        this.slideButton = ButtonPressState.UNPRESSED;
        this.slideManual = ButtonPressState.UNPRESSED;
        this.slideManualUp = ButtonPressState.UNPRESSED;
        this.ultimateButton = ButtonPressState.UNPRESSED;

        this.claw = new Claw(hardwareMap, telemetry);
        this.clawRotator = new ClawRotator(hardwareMap, telemetry);
    }

    @Override
    public void loop() {

        if (gamepad1.left_trigger>0.1f && slideButton == ButtonPressState.UNPRESSED) {
            slideButton = ButtonPressState.PRESSED_GOOD;
        } else if (gamepad1.left_trigger>0.1f && slideButton==ButtonPressState.PRESSED_GOOD) {
            slideButton = ButtonPressState.DEPRESSED;
        } else if (gamepad1.left_trigger<0.1f) slideButton = ButtonPressState.UNPRESSED;

        if (slides.getRotatorEncoder() < 600 && slides.getEncoder() > 200) {
            slides.retract();
        }

        if (gamepad1.left_bumper){
            slides.retract();
            telemetry.addLine("slides rotator: " + String.valueOf(slides.getRotatorEncoder()));
            telemetry.addLine("slides: " + String.valueOf(slides.getEncoder()));
            telemetry.update();
        } else if (gamepad1.right_bumper){
            if (slides.getRotatorEncoder() < 1000 && slides.getRotatorEncoder() > -1000) {
                stop();
            }
            else {
                slides.extend();
            }
            telemetry.addLine("slides rotator: " + String.valueOf(slides.getRotatorEncoder()));
            telemetry.addLine("slides: " + String.valueOf(slides.getEncoder()));
            telemetry.update();
        } else {
            slides.stop();
        }

//        DRIVETRAIN TELEMETRY
//        telemetry.addLine(String.valueOf(drive.getfl()) + "get front left");
//        telemetry.addLine(String.valueOf(drive.getbr()) + "get back right");
//        telemetry.addLine(String.valueOf(drive.getfr()) + "get front right");
//        telemetry.addLine(String.valueOf(drive.getbl()) + "get back left");
//        telemetry.update();

        //DRIVE
        float x = gamepad2.left_stick_x;
        float y = gamepad2.left_stick_y;
        float turn = gamepad2.right_stick_x;
        if (slowMode) {
            telemetry.addLine("slowmode");
            float slowx = (float)0.4*x;
            float slowy = (float)0.4*y;
            stick_margin = 0.3f;
            move(slowx, slowy, turn);
        } else {
            stick_margin = 0.1f;
            move(x, y, turn);
        }

        if (gamepad1.right_trigger > 0.5) {
            slides.rotateRight();
            telemetry.addLine("rotating slides: " + String.valueOf(slides.getRotatorEncoder()));

        } else if (gamepad1.left_trigger > 0.5){
            slides.rotateLeft();
            telemetry.addLine("rotating slides: " + String.valueOf(slides.getRotatorEncoder()));
            telemetry.update();
        } else {
            slides.stopRotator();
        }
        if (gamepad2.dpad_down){
            slowMode = true;
            telemetry.addLine("keep up: " + String.valueOf(slides.getRotatorEncoder()));
            telemetry.update();
        } else if (gamepad2.dpad_up){
            slowMode = false;
            telemetry.update();
        }

        // improved code by yours truly:
        if (gamepad1.y && (clawButton == ButtonPressState.UNPRESSED)) {
            if (!clawOpen) {
                claw.open();
                telemetry.addLine("open claw: " + claw.getPosition());
                telemetry.update();
                clawOpen = true;
                clawButton = ButtonPressState.PRESSED_GOOD;
            } else {
                claw.close();
                telemetry.addLine("close claw: " + claw.getPosition());
                telemetry.update();
                clawOpen = false;
                clawButton = ButtonPressState.PRESSED_GOOD;
            }

        } else if (!gamepad1.y && (clawButton == ButtonPressState.PRESSED_GOOD)) {
            clawButton = ButtonPressState.UNPRESSED;
        }







        if (gamepad1.a && (cRotatorButton == ButtonPressState.UNPRESSED) && !cRotatorAtDrop) {
            cRotatorButton = ButtonPressState.PRESSED_GOOD;
            clawRotator.toDrop();
            telemetry.addLine("claw to drop position: " + clawRotator.getPosition());
            cRotatorAtDrop = true;
        } else if (gamepad1.a && (cRotatorButton == ButtonPressState.UNPRESSED) && cRotatorAtDrop) {
            cRotatorButton = ButtonPressState.PRESSED_GOOD;
            clawRotator.toPick();
            telemetry.addLine("claw to pick position: " + clawRotator.getPosition());
            cRotatorAtDrop = false;
        } else if (!(gamepad1.a) && (cRotatorButton == ButtonPressState.PRESSED_GOOD)){
            cRotatorButton = ButtonPressState.UNPRESSED;
        }

        if ((gamepad2.left_trigger > 0.7f) && (ultimateButton == ButtonPressState.UNPRESSED)){
            try {
                hangSpecimen();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }

    }

    public void move(float x, float y, float turn) {
        // if the stick movement is negligible, set STICK_MARGIN to 0

        if (Math.abs(x) <= stick_margin) x = .0f;
        if (Math.abs(y) <= stick_margin) y = .0f;
        if (Math.abs(turn) <= stick_margin) turn = .0f;

        //Notation of a ? b : c means if a is true do b, else do c.
        double multiplier = normalPower;
        drive.move(x * multiplier, y * multiplier, -turn * multiplier);
    }

    public void setAutonAction (int action) {
        autonAction = action;
        runtime.reset();

    }
    public void hangSpecimen() throws InterruptedException {
        while(autonAction!=-1){
            switch (autonAction) {
                case 0: // move back
                    ultimateButton = ButtonPressState.PRESSED_GOOD;
                    while (runtime.seconds() < 3) {
                        telemetry.addLine("case0 " +  runtime.seconds());
                        telemetry.update();
                        drive.move(0, -1, 0);
                    }
                    setAutonAction(1);
                    telemetry.addLine(Integer.toString(autonAction));
                    break;
                case 1: // stop
                    telemetry.addLine("case1");
                    telemetry.update();
                    drive.move(0, 0, 0);
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(2);
                    }
                    break;
                case 2: // stop
                    telemetry.addLine("case2");
                    telemetry.update();
                    clawRotator.toDrop();
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(3);
                    }
                    break;
                case 3: // rotate slides
                    telemetry.addLine("case3");
                    telemetry.update();
                    slides.rotateRight();
                    if (runtime.seconds() > 0.8) {
                        runtime.reset();
                        setAutonAction(4);
                    }
                    break;
                case 4: // stop slides
                    telemetry.addLine("case4");
                    telemetry.update();
                    slides.stopRotator();
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(5);
                    }
                    break;
                case 5: // extend slides
                    telemetry.addLine("case5");
                    telemetry.update();
                    slides.extend();
                    if (runtime.seconds() > 0.8) {
                        runtime.reset();
                        setAutonAction(6);
                    }
                    break;
                case 6: // stop slides
                    telemetry.addLine("case6");
                    telemetry.update();
                    slides.stop();
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(7);
                    }
                    break;
                case 7: // rotate claws
                    telemetry.addLine("case7");
                    telemetry.update();
                    clawRotator.toPick();
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(8);
                    }
                    break;
                case 8: // slide retract a little bit
                    telemetry.addLine("case8");
                    telemetry.update();
                    slides.retract();
                    if (runtime.seconds() > 0.4) {
                        runtime.reset();
                        setAutonAction(9);
                    }
                    break;
                case 9: // stop slides
                    telemetry.addLine("case9");
                    telemetry.update();
                    slides.stop();
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(10);
                    }
                    break;
                case 10: // stop slides
                    telemetry.addLine("case10");
                    telemetry.update();
                    slides.stop();
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(11);
                    }
                    break;
                case 11: // drive forward a bit
                    telemetry.addLine("case11");
                    telemetry.update();
                    drive.move(0, 1, 0);
                    if (runtime.seconds() > 0.2) {
                        runtime.reset();
                        setAutonAction(12);
                    }
                    break;
                case 12: // stop driving
                    telemetry.addLine("case12");
                    telemetry.update();
                    drive.move(0, 0, 0);
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(13);
                    }
                    break;
                case 13: // let go
                    telemetry.addLine("case13");
                    telemetry.update();
                    claw.open();
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(14);
                    }
                    break;
                case 14: // rotate claws
                    telemetry.addLine("case14");
                    telemetry.update();
                    clawRotator.toDrop();
                    if (runtime.seconds() > 0.2) {
                        runtime.reset();
                        setAutonAction(15);
                    }
                    break;
                case 15: // retract slides
                    telemetry.addLine("case15");
                    telemetry.update();
                    slides.retract();
                    if (runtime.seconds() > 0.3) {
                        runtime.reset();
                        setAutonAction(16);
                    }
                    break;
                case 16: // top slides
                    telemetry.addLine("case16");
                    telemetry.update();
                    slides.stop();
                    if (runtime.seconds() > 0.1) {
                        runtime.reset();
                        setAutonAction(17);
                        ultimateButton = ButtonPressState.UNPRESSED;
                    }

                    break;
            }
        }
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/Teleop.kt
package org.firstinspires.ftc.teamcode.teleop

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.util.ElapsedTime
import org.firstinspires.ftc.teamcode.utilities.Claw
import org.firstinspires.ftc.teamcode.utilities.ClawRotator
import org.firstinspires.ftc.teamcode.utilities.Latch
import org.firstinspires.ftc.teamcode.utilities.PullUp
import org.firstinspires.ftc.teamcode.utilities.SimpleMecanumDrive
import org.firstinspires.ftc.teamcode.utilities.Slides
import kotlin.math.abs

//import org.firstinspires.ftc.teamcode.utilities.PullUp;
@Config
@TeleOp(name = "DriveTrain Teleop")
class Teleop : OpMode() {
    var slides: Slides? = null
    var pullup: PullUp? = null
    var claw: Claw? = null
    var latch: Latch? = null
    var clawRotator: ClawRotator? = null

    // ClawIteration2 claw;
    var drive: SimpleMecanumDrive? = null

    //Set normal power constant to 1, no point in slowing the robot down
    val normalPower: Double = 1.0

    // in case of joystick drift, ignore very small values
    val STICK_MARGIN: Float = 0.7f

    var clawState: Boolean = true
    var cRotatorAtDrop: Boolean = false

    //Prevents irreversible things such as pullup and plane launcher from running before this button is pressed
    // (will add later)
    // boolean validate = false;
    //makes validate button have to be pressed for a while before features enabled
    // int validatecount = 0;
    enum class SlideState {
        SLIDE_BOTTOM,
        SLIDE_LOW,
        SLIDE_MEDIUM,
        SLIDE_HIGH
    }

    enum class RotatorState {
        SLIDE_DOWN,
        SLIDE_UP,
        SLIDE_STOP
    }

    var slideState: SlideState = SlideState.SLIDE_BOTTOM
    var rotatorState: RotatorState = RotatorState.SLIDE_DOWN

    enum class PullUpState {
        NEUTRAL,
        REACH_UP
    }

    var pullupstate: PullUpState = PullUpState.REACH_UP

    var elapsedTime: ElapsedTime? = null

    enum class ButtonPressState {
        PRESSED_GOOD,  //the first time we see the button pressed
        DEPRESSED,  //you haven't let go
        UNPRESSED // its not pressed
    }

    var slideButton: ButtonPressState? = null
    var rotatorButton: ButtonPressState? = null
    var clawButton: ButtonPressState? = null
    var cRotatorButton: ButtonPressState? = null
    var slideManual: ButtonPressState? = null
    var slideManualUp: ButtonPressState? = null

    var started: Boolean = false

    // boolean intakeRunning = false;
    override fun init() {
        this.drive = SimpleMecanumDrive(hardwareMap)
        this.slides = Slides(hardwareMap, telemetry)
        this.pullup = PullUp(hardwareMap)
        this.clawButton = ButtonPressState.UNPRESSED
        this.cRotatorButton = ButtonPressState.UNPRESSED
        this.slideButton = ButtonPressState.UNPRESSED
        this.slideManual = ButtonPressState.UNPRESSED
        this.slideManualUp = ButtonPressState.UNPRESSED

        this.claw = Claw(hardwareMap, telemetry)
        this.latch = Latch(hardwareMap, telemetry)
        this.clawRotator = ClawRotator(hardwareMap, telemetry)

        // slides.startPosition();
    }

    override fun loop() {
        //VALIDATE
//        if (gamepad1.dpad_up) {++validatecount;}
//        if (validatecount > 5) {validate = true;}

        if (gamepad1.left_trigger > 0.1f && slideButton == ButtonPressState.UNPRESSED) {
            slideButton = ButtonPressState.PRESSED_GOOD
        } else if (gamepad1.left_trigger > 0.1f && slideButton == ButtonPressState.PRESSED_GOOD) {
            slideButton = ButtonPressState.DEPRESSED
        } else if (gamepad1.left_trigger < 0.1f) slideButton = ButtonPressState.UNPRESSED

        //        if (gamepad1.right_bumper && slideManual == ButtonPressState.UNPRESSED) {
//            slideManual = ButtonPressState.PRESSED_GOOD;
//        } else if (gamepad1.right_bumper && slideManual==ButtonPressState.PRESSED_GOOD) {
//            slideManual = ButtonPressState.DEPRESSED;
//        } else if (!gamepad1.right_bumper) slideManual = ButtonPressState.UNPRESSED;
//
//        if (gamepad1.left_bumper && slideManualUp == ButtonPressState.UNPRESSED) {
//            slideManualUp = ButtonPressState.PRESSED_GOOD;
//        } else if (gamepad1.left_bumper && slideManualUp==ButtonPressState.PRESSED_GOOD) {
//            slideManualUp = ButtonPressState.DEPRESSED;
//        } else if (!gamepad1.left_bumper) slideManualUp = ButtonPressState.UNPRESSED;
        if (gamepad1.left_bumper) { //slideManual==ButtonPressState.PRESSED_GOOD) {
            slides!!.retract()
            telemetry.addLine("retracting")
        } else if (gamepad1.right_bumper) { //slideManualUp==ButtonPressState.PRESSED_GOOD) {
            slides!!.extend()
        } else {
            slides!!.stop()
        }

        //DRIVE
        val x = gamepad1.left_stick_x
        val y = gamepad1.left_stick_y
        val turn = gamepad1.right_stick_x
        move(x, -y, turn)

        when (slideState) {
            SlideState.SLIDE_BOTTOM -> if (abs((slides.getEncoder() - 30).toDouble()) < 10) { // drop height
                if (slideButton == ButtonPressState.PRESSED_GOOD) {
                    slideState = SlideState.SLIDE_HIGH
                    telemetry.addData("pos", slides.getEncoder())
                    telemetry.update()
                    // slides.high();
                }
            }

            SlideState.SLIDE_LOW -> {
                telemetry.addData(
                    "u shouldnt see this lol but encoder position",
                    slides.getEncoder()
                )
                telemetry.update()
            }

            SlideState.SLIDE_HIGH -> if (abs((slides.getEncoder() - 30).toDouble()) < 10) { // high height 3481
                if (slideButton == ButtonPressState.PRESSED_GOOD) {
                    slides!!.tozero()
                    slideState = SlideState.SLIDE_BOTTOM
                    telemetry.addData("pos", slides.getEncoder())
                    telemetry.update()
                }
            }

            else -> {
                slideState = SlideState.SLIDE_BOTTOM
                telemetry.addLine("default")
                telemetry.update()
            }
        }
        when (pullupstate) {
            PullUpState.NEUTRAL -> if (gamepad1.x) {
                pullup!!.manualLeftUp()
                pullup!!.manualRightUp()
                pullupstate = PullUpState.REACH_UP
            }

            PullUpState.REACH_UP -> {
                telemetry.addData("pullup1pos", +pullup.getPosition1())
                telemetry.addData("pullup2pos", +pullup.getPosition2())
                telemetry.update()
                if (gamepad1.x) {
                    pullup!!.manualLeftDown()
                    pullup!!.manualRightDown()
                    pullupstate = PullUpState.NEUTRAL
                }
            }

            else -> pullupstate = PullUpState.NEUTRAL
        }
        if (gamepad1.dpad_left) {
            latch!!.latchOn()
        }

        if (gamepad1.dpad_right) {
            latch!!.latchOff()
        }


        //  ROTATING SLIDES
        if (gamepad1.right_trigger > 0.5) {
            slides!!.rotateRight()
            telemetry.addLine("rotator going up")
            telemetry.update()
        } else if (gamepad1.left_trigger > 0.5) {
            slides!!.rotateLeft()
            telemetry.addLine("rotator going down")
            telemetry.update()
        } else {
            if (slides.getRotatorEncoder() >= 400) {
                slides!!.keepUp()
                telemetry.addLine("kept up")
                telemetry.update()
            } else {
                slides!!.stopRotator()
            }
        }

        /* previous code
        if (gamepad1.b && (clawButton == ButtonPressState.UNPRESSED) && !clawState) {
            clawButton = ButtonPressState.PRESSED_GOOD;
            claw.open(
            );
            clawState = true;


        } else if (gamepad1.x && (clawButton == ButtonPressState.UNPRESSED) && clawState) {
            clawButton = ButtonPressState.PRESSED_GOOD;
            claw.close();
            clawState = false;
        } else if (!(gamepad1.b) && (!(gamepad1.x) && (clawButton == ButtonPressState.PRESSED_GOOD))){
            clawButton = ButtonPressState.UNPRESSED;

        }
         */

        // improved code by yours truly:
        if (gamepad1.y && (clawButton == ButtonPressState.UNPRESSED)) {
            if (!clawState) {
                claw!!.open()
                clawState = true
                clawButton = ButtonPressState.PRESSED_GOOD
            } else {
                claw!!.close()
                clawState = false
                clawButton = ButtonPressState.PRESSED_GOOD
            }
        } else if (!gamepad1.y && (clawButton == ButtonPressState.PRESSED_GOOD)) {
            clawButton = ButtonPressState.UNPRESSED
        }


        if (gamepad1.a && (cRotatorButton == ButtonPressState.UNPRESSED) && !cRotatorAtDrop) {
            cRotatorButton = ButtonPressState.PRESSED_GOOD
            clawRotator!!.toDrop()
            cRotatorAtDrop = true
        } else if (gamepad1.a && (cRotatorButton == ButtonPressState.UNPRESSED) && cRotatorAtDrop) {
            cRotatorButton = ButtonPressState.PRESSED_GOOD
            clawRotator!!.toPick()
            cRotatorAtDrop = false
        } else if (!(gamepad1.a) && (cRotatorButton == ButtonPressState.PRESSED_GOOD)) {
            cRotatorButton = ButtonPressState.UNPRESSED
        }

        //        if (gamepad1.y) {
//            clawRotator.toDrop();
//            cRotatorAtDrop = true;
//        } else if (gamepad1.a) {
//            clawRotator.toPick();
//            cRotatorAtDrop = false;
//        }
        // THE ULTIMATE BUTTON
        if (gamepad1.dpad_up) {
            //retract slides
            //rotate slides
            //latch on
            //extend slides

            val slidesRunnable = Runnable {
                while (slides.getEncoder() <= -100) {
                    slides!!.retract()
                }
                //TODO: tune
                while (slides.getRotatorEncoder() <= 470) {
                    slides!!.rotateRight()
                }

                latch!!.latchOn()
                while (slides.getEncoder() >= -3000) {
                    slides!!.extend()
                }
            }
            val slidesThread = Thread(slidesRunnable)
            slidesThread.start()
        } else if (gamepad1.dpad_down) {
            //retract slides
            //rotate slides
            //latch off
            //extend slides

            val slidesRunnable = Runnable {
                while (slides.getEncoder() <= -100) {
                    slides!!.retract()
                }
                //TODO: tune
                while (slides.getRotatorEncoder() >= 200) {
                    slides!!.rotateLeft()
                }

                latch!!.latchOn()
                while (slides.getEncoder() >= -300) {
                    slides!!.extend()
                }
            }
            val slidesThread = Thread(slidesRunnable)
            slidesThread.start()
        }
    }

    fun move(x: Float, y: Float, turn: Float) {
        // if the stick movement is negligible, set STICK_MARGIN to 0
        var x = x
        var y = y
        var turn = turn
        if (abs(x.toDouble()) <= STICK_MARGIN) x = .0f
        if (abs(y.toDouble()) <= STICK_MARGIN) y = .0f
        if (abs(turn.toDouble()) <= STICK_MARGIN) turn = .0f

        //Notation of a ? b : c means if a is true do b, else do c.
        val multiplier = normalPower
        drive!!.move(-x * multiplier, y * multiplier, turn * multiplier)
    }
}
